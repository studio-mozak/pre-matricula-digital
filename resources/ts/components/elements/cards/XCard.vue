<script lang="ts">
import { computed, defineComponent, h } from 'vue';
import { hSlot } from '@/utils/render';
export default defineComponent({
  name: 'XCard',
  props: {
    tag: {
      type: String,
      default: 'div',
    },
    square: Boolean,
    flat: Boolean,
    bordered: Boolean,
    hoverable: Boolean,
  },
  setup(props, { slots }) {
    /* eslint-disable prefer-template */
    const classes = computed(
      () =>
        'x-card' +
        (props.bordered === true ? ' x-card--bordered' : '') +
        (props.hoverable === true ? ' x-card--hover' : '') +
        (props.square === true ? ' x-card--square no-border-radius' : '') +
        (props.flat === true ? ' x-card--flat no-shadow' : '')
    );
    return () => h(props.tag, { class: classes.value }, hSlot(slots));
  },
});
</script>
