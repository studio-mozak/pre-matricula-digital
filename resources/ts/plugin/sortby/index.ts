/* eslint-disable */
export {};

declare global {
  interface Array<T> {
    sortBy(key: string | number): Array<T>;
  }
}

if (!Array.prototype.sortBy) {
  Array.prototype.sortBy = function <T>(key: string | number): T[] {
    return this.sort((a: any, b: any) => (a[key] > b[key] ? 1 : -1));
  };
}
