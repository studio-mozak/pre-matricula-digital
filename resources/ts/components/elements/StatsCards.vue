<template>
  <div v-if="stats" class="row mt-5 mb-5 pl-lg-0 pl-3 pr-lg-0 pr-3">
    <div
      class="col-12 p-0 d-flex justify-content-lg-around"
      style="overflow-x: auto"
    >
      <x-card
        v-for="(card, index) in statsCards"
        :key="index + '-stats-cards'"
        :class="card.cardClass"
        bordered
        class="bg-background card-summary mr-3 mr-lg-0"
      >
        <x-card-section class="d-flex justify-content-center mt-3 mb-3 pb-0">
          <div
            class="rounded-circle p-3 border-card"
            :class="card.bgColor"
          ></div>
        </x-card-section>
        <x-card-section class="text-center pt-0 pb-3">
          <p class="card-number">
            {{ card.quantity }}
          </p>
          <p class="card-text">
            {{ card.text }}
          </p>
        </x-card-section>
      </x-card>
    </div>
  </div>
  <div v-else class="row mt-5 mb-5 pl-lg-0 pl-3 pr-lg-0 pr-3">
    <div
      class="col-12 p-0 d-flex justify-content-lg-around"
      style="overflow-x: auto"
    >
      <skeleton-card-stats
        v-for="(_, index) in 5"
        :key="index + '-stats-cards-skeleton'"
      />
    </div>
  </div>
</template>

<script setup lang="ts">
import SkeletonCardStats from '@/components/loaders/components/CardStats.vue';
import { StatProcesses } from '@/modules/preregistration/types';
import XCard from '@/components/elements/cards/XCard.vue';
import XCardSection from '@/components/elements/cards/XCardSection.vue';
import { computed } from 'vue';

const props = defineProps<{
  stats?: StatProcesses;
}>();

const statsCards = computed(() => {
  if (!props.stats || Object.keys(props.stats as StatProcesses).length === 0)
    return [];

  return [
    {
      quantity: props.stats ? props.stats.vacancies : null,
      text: 'Vagas ofertadas',
      bgColor: 'bg-blue svg-real-estate',
      cardClass: 'col-6 col-md-2',
    },
    {
      quantity: props.stats ? props.stats.total : null,
      text: 'Pré-matrículas inscritas',
      bgColor: 'bg-blue svg-perfis',
      cardClass: 'col-6 col-md-2',
    },
    {
      quantity: props.stats ? props.stats.accepted : null,
      text: 'Deferidas',
      bgColor: 'bg-green svg-aprovado',
      cardClass: 'col-6 col-md-2',
    },
    {
      quantity: props.stats ? props.stats.rejected : null,
      text: 'Indeferidas',
      bgColor: 'bg-red svg-desaprovado',
      cardClass: 'col-6 col-md-2',
    },
    {
      quantity: props.stats ? props.stats.waiting : null,
      text: 'Em espera',
      bgColor: 'bg-yellow svg-esperando',
      cardClass: 'col-6 col-md-2',
    },
  ];
});
</script>
