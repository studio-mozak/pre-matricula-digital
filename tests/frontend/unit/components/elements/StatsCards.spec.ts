import '@/plugin/sortby';
import {
  StatProcesses,
  StatProcessesProps,
} from '@/modules/preregistration/types';

import { VueWrapper, flushPromises, shallowMount } from '@vue/test-utils';
import StatsCards from '@/components/elements/StatsCards.vue';
import XCard from '@/components/elements/cards/XCard.vue';

import { expect, vi } from 'vitest';

const propsMock: StatProcessesProps = {
  stats: {
    accepted: 3197,
    rejected: 1227,
    total: 4890,
    vacancies: 25518,
    waiting: 466,
  },
};

const statesCardsMock = [
  {
    quantity: 25518,
    text: 'Vagas ofertadas',
    bgColor: 'bg-blue svg-real-estate',
    cardClass: 'col-6 col-md-2',
  },
  {
    quantity: 4890,
    text: 'Pré-matrículas inscritas',
    bgColor: 'bg-blue svg-perfis',
    cardClass: 'col-6 col-md-2',
  },
  {
    quantity: 3197,
    text: 'Deferidas',
    bgColor: 'bg-green svg-aprovado',
    cardClass: 'col-6 col-md-2',
  },
  {
    quantity: 1227,
    text: 'Indeferidas',
    bgColor: 'bg-red svg-desaprovado',
    cardClass: 'col-6 col-md-2',
  },
  {
    quantity: 466,
    text: 'Em espera',
    bgColor: 'bg-yellow svg-esperando',
    cardClass: 'col-6 col-md-2',
  },
];

const onMounted = async (props = propsMock) => {
  const wrapper = shallowMount(StatsCards, {
    props,
  });

  await wrapper.vm.$nextTick();
  await flushPromises();

  return wrapper;
};

describe('StatsCards', () => {
  let wrapper: VueWrapper<InstanceType<typeof StatsCards>>;

  beforeEach(async () => {
    wrapper = await onMounted();

    await wrapper.vm.$nextTick();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('the component exists', () => {
    expect(wrapper).toBeTruthy();
  });

  test('the initial variables must be loaded with the correct data', () => {
    expect(wrapper.vm.stats).toMatchObject(propsMock.stats);
    expect(wrapper.vm.statsCards).toMatchObject(statesCardsMock);
  });

  test('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('the component must render the correct data', () => {
    const component = wrapper.findAllComponents({
      name: 'x-card',
    });

    expect(component).toBeTruthy();
    expect(component.length).toBe(5);
  });

  test('if process is empty, the component must not render', async () => {
    wrapper = await onMounted({
      stats: {} as StatProcesses,
    });

    const component = wrapper.findAllComponents({
      name: 'x-card',
    });

    await wrapper.vm.$nextTick();
    await flushPromises();

    expect(wrapper.vm.stats).toMatchObject({});
    expect(wrapper.vm.statsCards).toMatchObject([]);
    expect(component.length).toBe(0);
  });
});
