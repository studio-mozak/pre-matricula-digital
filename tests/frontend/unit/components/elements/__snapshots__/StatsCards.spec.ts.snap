// Vitest Snapshot v1

exports[`StatsCards > should match snapshot 1`] = `
<div
  class="row mt-5 mb-5 pl-lg-0 pl-3 pr-lg-0 pr-3"
>
  <div
    class="col-12 p-0 d-flex justify-content-lg-around"
    style="overflow-x: auto;"
  >
    
    <x-card-stub
      bordered="true"
      class="col-6 col-md-2 bg-background card-summary mr-3 mr-lg-0"
      flat="false"
      hoverable="false"
      square="false"
      tag="div"
    />
    <x-card-stub
      bordered="true"
      class="col-6 col-md-2 bg-background card-summary mr-3 mr-lg-0"
      flat="false"
      hoverable="false"
      square="false"
      tag="div"
    />
    <x-card-stub
      bordered="true"
      class="col-6 col-md-2 bg-background card-summary mr-3 mr-lg-0"
      flat="false"
      hoverable="false"
      square="false"
      tag="div"
    />
    <x-card-stub
      bordered="true"
      class="col-6 col-md-2 bg-background card-summary mr-3 mr-lg-0"
      flat="false"
      hoverable="false"
      square="false"
      tag="div"
    />
    <x-card-stub
      bordered="true"
      class="col-6 col-md-2 bg-background card-summary mr-3 mr-lg-0"
      flat="false"
      hoverable="false"
      square="false"
      tag="div"
    />
    
  </div>
</div>
`;
